syntax = "proto3";

package logistics;

// ------------------------------
// 资源定义
// ------------------------------
message Item {
  // 货号（资源唯一标识）
  string sku = 1;
  string name = 2;
  string description = 3;
  int64 quantity = 4;
}

// ------------------------------
// 服务接口
// ------------------------------
service InventoryService {
  // 1) Add Item Resource
  rpc AddItem(AddItemRequest) returns (AddItemResponse);

  // 2) Update Item Resource
  rpc UpdateItem(UpdateItemRequest) returns (UpdateItemResponse);

  // 3) Take Item Resource（出库/扣减）
  rpc TakeItem(TakeItemRequest) returns (TakeItemResponse);

  // 4) Query Item Resource
  rpc QueryItem(QueryItemRequest) returns (QueryItemResponse);
}

// 日志服务
service LoggerService {
  rpc LogOperation(LogRequest) returns (LogResponse);
  rpc QueryLogs(QueryLogsRequest) returns (QueryLogsResponse);
  rpc GetStats(StatsRequest) returns (StatsResponse);
  rpc ClearLogs(ClearLogsRequest) returns (ClearLogsResponse);
}

// ------------------------------
// Add
// ------------------------------
message AddItemRequest {
  Item item = 1;
}

message AddItemResponse {
  Item item = 1;
}

// ------------------------------
// Update
// ------------------------------
message UpdateItemRequest {
  string sku = 1;   // 要更新的货号
  string name = 2;
  string description = 3;
  int64 quantity = 4;
}

message UpdateItemResponse {
  Item item = 1;
}

// ------------------------------
// Take
// ------------------------------
message TakeItemRequest {
  string sku = 1;   // 被扣减的货号
  int64 amount = 2; // 扣减数量
}

message TakeItemResponse {
  Item item = 1;    // 扣减后最新状态
}

// ------------------------------
// Query
// ------------------------------
message QueryItemRequest {
  string sku = 1;
}

message QueryItemResponse {
  Item item = 1;
}

// ==================== 日志服务消息定义 ====================

// 日志记录请求
message LogRequest {
  string service_name = 1;      // 服务名称
  string operation = 2;         // 操作类型
  string client_ip = 3;         // 客户端IP
  bool success = 4;             // 操作是否成功
  string request_data = 5;      // 请求数据（JSON字符串）
  string response_data = 6;     // 响应数据（JSON字符串）
  string error_message = 7;     // 错误信息（可选）
}

// 日志记录响应
message LogResponse {
  bool success = 1;
  string message = 2;
}

// 查询日志请求
message QueryLogsRequest {
  string service_name = 1;      // 服务名称过滤（可选）
  string operation = 2;         // 操作类型过滤（可选）
  int32 limit = 3;              // 返回记录数量限制
}

// 查询日志响应
message QueryLogsResponse {
  repeated LogEntry logs = 1;
  int32 total_count = 2;
}

// 日志条目
message LogEntry {
  string timestamp = 1;
  string service_name = 2;
  string operation = 3;
  string client_ip = 4;
  bool success = 5;
  string request_data = 6;
  string response_data = 7;
  string error_message = 8;
}

// 统计信息请求
message StatsRequest {
  // 空请求，返回所有统计信息
}

// 统计信息响应
message StatsResponse {
  int32 total_operations = 1;
  int32 successful_operations = 2;
  int32 failed_operations = 3;
  double success_rate = 4;
  repeated ServiceStats service_stats = 5;
  repeated OperationStats operation_stats = 6;
}

// 服务统计
message ServiceStats {
  string service_name = 1;
  int32 total = 2;
  int32 success = 3;
  int32 failed = 4;
  double success_rate = 5;
}

// 操作统计
message OperationStats {
  string operation = 1;
  int32 total = 2;
  int32 success = 3;
  int32 failed = 4;
  double success_rate = 5;
}

// 清空日志请求
message ClearLogsRequest {
  // 空消息，不需要参数
}

// 清空日志响应
message ClearLogsResponse {
  bool success = 1;
  string message = 2;
  int32 cleared_count = 3;
}